### PR TITLE
iox-#2052 Refactor copy_and_move_impl

### LINKED ISSUE #2052: Implement copy/move constructors and assignment operators for 'FixedPositionContainer'
## Brief feature description

The `FixedPositionContainer` is neither move nor copy constructible or assignable. Currently it is not required but would be a nice enhancement.

### PR DESCRIPTION
**Pull request by Dennis40816**
Refactor `FixedPositionContainer::copy_and_move_impl` by new struct `SpecialCreationHelper` and enum class `SpecialCreationOperations` with C++17 features.

## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

I implement a prototype named `SpecialCreationHelper` according to  [previous discussion](https://github.com/eclipse-iceoryx/iceoryx/pull/2069#issuecomment-1804246731).
If it looks good to you, I'll create another issue to move the helper class to `design` module.

Btw, I have no idea if `move_or_copy` and `move_or_copy_it` are needed or not. These functions currently cause twice time `moveCtor` calls problem (See [example](https://godbolt.org/z/o1TW3866Y)). Currently, I directly use `if constexpr` to pass correct value. If you think they are unnecessary, I will delete related comment and code.
 
## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References
- #2069 
- Closes #2052

### CODE REVIEW

**Comment by Dennis40816** [comment]
@elBoberido Please review this when you have a moment. Thank you :)

**Comment by elBoberido** [comment]
@Dennis40816 it's on my todo list for today. I hope I find some time :)

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (elBoberido) [inline iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:110]: Please also add a small comment why this is necessary. Something simple like `status array is not yet initialized`

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (elBoberido) [inline iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:124]: I'm not sure I understand the comment :sweat_smile: 

Is this something temporal which will be fixed in this PR or is it just something which could replace the lines below?

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (elBoberido) [inline iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:97]: I wonder whether there is a nice way to move this to `SpecialCreationHelper`?

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (elBoberido) [inline iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:28]: I'm thinking about another name. I guess many developer won't immediately understand what this is for, including me if I would not have reviewed the last PR and potentially also my future self in a year again.

What do you think about replacing `SpecialCreation...` with `MoveAndCopy...`

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (elBoberido) [inline iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:49]: This name is also a bit misleading since it may also assign. Maybe we find a more fitting name.

**Review by elBoberido — COMMENTED** [review]
Nice. Looks good overall. I think this could be moved to `iceoryx_hoofs/design` in a follow up PR. Then the discussion about names might get hot :)

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:124]: Ah, I should have specified that it's the `move_or_copy_it` issue. My intention was to replace the code from lines 126 to 133 with that on line 124, but I was unsuccessful. In the case of `movector`, it is called once within `move_or_copy_it`, which is not our desired outcome. Therefore, I'm still searching for a solution to this problem. You can view an example [here](https://godbolt.org/z/3xcPraPTh). If you have any suggestions, please feel free to share them with me.

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:28]: Good, it's more clear

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:97]: I think this can be remove directly, since we also check this properties in `ctor_create` and `assignment_create`. What do you think :)

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:49]: Indeed, finding a single term that encapsulates both construction and assignment is quite challenging :(

While `construct` seems like a suitable name, it could potentially mislead users into thinking the function only deals with object construction. I initially chose `create` precisely because it broadly suggests the creation of a new instance, encompassing both construction and assignment functionalities.

Do you have any idea?

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (elBoberido) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:124]: Okay. Would you move the commented code out before the merge if you don't find a solution. You could for example add a section to the description of #2052 and add the code there. Something like `## Evaluate foo` with an explanation or create a new issue ... might even be the issue to move the helper files to `iceoryx_hoofs/design`

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (elBoberido) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:97]: Yes, remove it. No need to do the check multiple times.

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (elBoberido) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:49]: Not really. The only thing I can think of is `copy_or_move`. This shouldn't block the PR though. We can take care of the name when the helper struct is moved to `iceoryx_hoofs/design`

File: iceoryx_dust/container/include/iox/detail/fixed_position_container.inl
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container.inl:124]: Ok.

File: iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp
Comment (Dennis40816) [reply iceoryx_dust/container/include/iox/detail/fixed_position_container_helper.hpp:49]: I'll use `transfer` as a temp name.

**Review by elBoberido — APPROVED** [review]
Great job. Like already mentioned, the helper struct could be reused by other classes and it would be great if you had some time to move it to `iceoryx_hoofs/design` and write a small design document in `doc/design`. Just copy `_template.md`. If a sections does not apply just delete it.

I think it could be done as part of #1686. If you want to test the abstraction on another class then #1517 would be ideal :)

**Comment by Dennis40816** [comment]
No problem. I'll create an new issue about moving it to `iceoryx_hoofs/design` this weekend!

**Pull request merged**
